pkg/build/controller/strategy/util.go

```diff
@@ -649,3 +662,47 @@ func setupBlobCache(pod *corev1.Pod) {
 		pod.Spec.InitContainers = initContainers
 	}
 }
+
+// setupBuildVolumes sets up user defined BuildVolumes
+func setupBuildVolumes(pod *corev1.Pod, buildVolumes []buildv1.BuildVolume) error {
+	// if there are no BuildVolumes or the pod is nil,
+	// there is no processing needed, so just return quickly
+	if len(buildVolumes) == 0 || pod == nil {
+		return nil
+	}
+
+	usedUserVolumeMounts := make(map[string]struct{})
+
+	// iterate over existing VolumeMounts and add to the map
+	if len(pod.Spec.Containers) != 0 {
+		for _, vm := range pod.Spec.Containers[0].VolumeMounts {
+			usedUserVolumeMounts[vm.MountPath] = struct{}{}
+		}
+	}
+
+	for _, buildVolume := range buildVolumes {
+		// check for user provided mountPath collisions and return an error if one is found
+		for _, bvm := range buildVolume.Mounts {
+			if _, ok := usedUserVolumeMounts[bvm.DestinationPath]; ok {
+				// fail if a collision is found
+				return fmt.Errorf("user provided BuildVolumeMount path %q collides with VolumeMount path created by the build controller", bvm.DestinationPath)
+			}
+		}
+
+		volumeSource := corev1.VolumeSource{}
+
+		// if no collisions are found go ahead and add the volume/volume mount to the pod/container
+		switch {
```

Comment:
can't this be

```suggestion
		switch buildVolume.Source.Type {
```

(then drop the equality checks in the cases)

Reply:
updated in next push. I was thinking of another language.